### PR TITLE
WIP - Kindly don't merge to main - 7600 end to end monitoring ess api

### PR DESCRIPTION
**Pull request by Avani13709**
code changes

### CODE REVIEW

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs [hunk 1]
```diff
@@ -159,9 +185,12 @@ public async Task<bool> CreateCatalogFile(string batchId, string exchangeSetRoot
 
             if (!string.IsNullOrWhiteSpace(exchangeSetRootPath))
             {
+                DateTime createCatalogueFileTaskStartAt = DateTime.UtcNow;
                 logger.LogInformation(EventIds.CreateCatalogFileRequestStart.ToEventId(), "Create catalog file request started for BatchId:{BatchId} and _X-Correlation-ID:{CorrelationId}", batchId, correlationId);
                 isFileCreated = await fulfilmentAncillaryFiles.CreateCatalogFile(batchId, exchangeSetRootPath, correlationId, listFulfilmentData, salesCatalogueDataResponse);
                 logger.LogInformation(EventIds.CreateCatalogFileRequestCompleted.ToEventId(), "Create catalog file request completed for BatchId:{BatchId} and _X-Correlation-ID:{CorrelationId}", batchId, correlationId);
+                DateTime createCatalogueFileTaskCompletedAt = DateTime.UtcNow;
+                monitorHelper.MonitorRequest("Create Catalogue File Task", createCatalogueFileTaskStartAt, createCatalogueFileTaskCompletedAt, correlationId,null,null, null,batchId);
```
Comment (janvi13603) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs:193]: Kindly rename Create Catalogue File Task to "Create Catalog File Task" and createCatalogueFileTaskStartAt to createCatalogFileTaskStartedAt and createCatalogueFileTaskCompletedAt to createCatalogFileTaskCompletedAt

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Models/FileShareService/Response/FulfilmentDataResponse.cs [hunk 2]
```diff
@@ -15,5 +15,7 @@ public class FulfilmentDataResponse
         public IEnumerable<string> FileUri { get; set; }
 
         public IEnumerable<BatchFile> Files { get; set; }
+        
+        public int TotalHitCountForQueryFileShareService { get; set; }
```
Comment (janvi13603) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Models/FileShareService/Response/FulfilmentDataResponse.cs:19]: Kindly change variable name from TotalHitCountForQueryFileShareService to FileShareServiceSearchQueryCount to maintain consistency

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs [hunk 3]
```diff
@@ -43,34 +45,54 @@ public class FulfilmentDataService : IFulfilmentDataService
...
         public async Task<string> CreateExchangeSet(SalesCatalogueServiceResponseQueueMessage message)
         {
+            DateTime createExchangeSetTaskStartAt = DateTime.UtcNow;
             string homeDirectoryPath = configuration["HOME"];
             var exchangeSetPath = Path.Combine(homeDirectoryPath, DateTime.UtcNow.ToString("ddMMMyyyy"), message.BatchId, fileShareServiceConfig.Value.ExchangeSetFileFolder);
             var exchangeSetRootPath = Path.Combine(exchangeSetPath, fileShareServiceConfig.Value.EncRoot);
             var exchangeSetPathForUploadZipFile = Path.Combine(homeDirectoryPath, DateTime.UtcNow.ToString("ddMMMyyyy"), message.BatchId);
             var listFulfilmentData = new List<FulfilmentDataResponse>();
-
+            DateTime downloadEncFilesFromFileShareServiceTaskStartAt = DateTime.UtcNow;
```
Comment (janvi13603) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs:59]: This line of code has to be inside of if statement at line 61 and also change variable name downloadEncFilesFromFileShareServiceTaskStartAt to queryAndDownloadEncFilesFromFileShareServiceTaskStartedAt

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs [hunk 4]
```diff
@@ -43,34 +45,54 @@ public class FulfilmentDataService : IFulfilmentDataService
...
+                    {
+                        totalHitCountsToFileShareServiceForQuery += queryCount;
+                    }
+                    listFulfilmentData.AddRange(fulfilmentDataResponses);
                 });
-                await Task.WhenAll(tasks);
+
+                await Task.WhenAll(tasks);                
+                DateTime downloadEncFilesFromFileShareServiceTaskCompletedAt = DateTime.UtcNow;
+                int folderCount = 0;
```
Comment (janvi13603) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs:82]: Rename variable totalHitCountsToFileShareServiceForQuery at line 75 from to 
fileShareServiceSearchQueryCount

and variable folderCount at line 82 to downloadedENCFileCount

Also rename "Query and download file share service data" at line 87 to "Query and Download ENC Files Task"

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/MonitorHelper.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        private const string requestStartedAt = "StartedAt";
+        private const string requestCompletedAt = "CompletedAt";
+        private const string runtimeDurationInMs = "DurationInMilliSeconds";
+        public MonitorHelper(IConfiguration configuration, TelemetryConfiguration telemetryConfiguration)
+        {
+            this.configuration = configuration;
+            this.telemetryClient = new TelemetryClient(telemetryConfiguration);
+        }
+
+        public void MonitorRequest(string message,DateTime startAt, DateTime completedAt, string correlationId, int? totalHitCountsToFileShareServiceForQuery = null, int? fileCount=null, long? fileSize= null,string batchId = null)
```
[the remark this answers is not in the document]
Comment (Avani13709) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/MonitorHelper.cs]: Done

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs [hunk 6]
```diff
@@ -139,11 +164,12 @@ public async Task CreateSerialEncFile(string batchId, string exchangeSetPath, st
...
         {
             bool isZipFileUploaded = false;
-
+            DateTime createZipFileTaskStartAt = DateTime.UtcNow;
             logger.LogInformation(EventIds.CreateZipFileRequestStart.ToEventId(), "Started creating exchange set zip file for BatchId:{BatchId} and _X-Correlation-ID:{CorrelationId}", batchId, correlationId);
             bool isZipFileCreated = fulfilmentFileShareService.CreateZipFileForExchangeSet(batchId, exchangeSetPath, correlationId);
             logger.LogInformation(EventIds.CreateZipFileRequestCompleted.ToEventId(), "Ended creating exchange set zip file for BatchId:{BatchId} and _X-Correlation-ID:{CorrelationId}", batchId, correlationId);
-
+            DateTime createZipFileTaskCompletedAt = DateTime.UtcNow;
+            monitorHelper.MonitorRequest("Zip File Creation Task", createZipFileTaskStartAt, createZipFileTaskCompletedAt, correlationId,null,null, null,batchId);
```
[the remark this answers is not in the document]
Comment (Avani13709) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs:172]: Done

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs (hunk 1 above)
Comment (Avani13709) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs]: Done

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs (hunk 3 above)
Comment (Avani13709) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs]: Done

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs (hunk 4 above)
Comment (Avani13709) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs]: Done

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Models/FileShareService/Response/FulfilmentDataResponse.cs (hunk 2 above)
Comment (Avani13709) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Models/FileShareService/Response/FulfilmentDataResponse.cs]: Done

**Review by janvi13603 — APPROVED** [review]
Thanks for fixing review comments.

**Review by martyn-fewtrell — APPROVED** [review]
Looks good

**Comment by Avani13709** [comment]
Closing this pull request as new PR has been raised

**Pull request closed without merging**